### DISCUSSION TITLE
How to create machining feature primitives?

### DISCUSSION THREAD

**Discussion by Boehm92** [jimy-byerley/pymadcad#74] [Q&A] [answered]
Hi all, 

i am currently trying to develop a machine learning framework for machining feature recognition like Zhang et al. presented  with the FeatureNet Framework (https://github.com/zibozzb/FeatureNet). They introduced 24 different machining features applied to 10mm cubes in .stl format. 
![IFeatureNet](https://user-images.githubusercontent.com/40231115/230732986-918a98bd-840c-40cb-bfd5-404879af8001.png)
The FeatureNet training data was programmed using the SolidWorks API. However, Solid Works is not free and FeatureNet has not published the API code.
So i programed a data generator with pymadcad to generate automatically training data of the 24 machining features. My machine learning model has a very high accuracy on the pymadcad data. However, when I try to test the machine learning model on the FeatureNet dataset, i have very poor accuracies on isolated machining features.  Due to the fact, that the model works well on 70% of all the other features, i think that the mesh topology of those isolated features differs in some kind from the solid works meshes.

One example is the "circular through slot"
FeatureNet:
![image](https://user-images.githubusercontent.com/40231115/230733487-ad984c31-557d-4efc-9121-b56e8daf9833.png)

PyMadCad:
![image](https://user-images.githubusercontent.com/40231115/230733953-b0994f9f-5534-4cd6-9d66-ae1fcadec23e.png)

I create the trough slot with following code:
```

import madcad as mdc
import numpy as np

cube = mdc.brick(width=mdc.vec3(10))
cube = cube.transform(mdc.vec3(5, 5, 5))

R = np.random.uniform(1, 4.5)
A = mdc.vec3(R, -0.01, 10.01)  # the .01 are important, so the difference operations can build a clear mesh
B = mdc.vec3(R / 2, -0.01, 10.01 - R / 2)
C = mdc.vec3(0, -0.01, 10.01)
X = np.random.uniform(0.5, 9.5 - R)

_lines = [mdc.Segment(A, C), mdc.ArcThrough(C, B, A)]
circular_trough_slot_web = mdc.web(_lines)
_circular_trough_slot = mdc.extrusion(10.02 * mdc.Y, mdc.flatsurface(circular_trough_slot_web))
_circular_trough_slot = _circular_trough_slot.transform(mdc.translate(mdc.vec3(X, 0, 0)))

updated_model = mdc.difference(cube, _circular_trough_slot)
updated_model.mergeclose()

mdc.show([updated_model])
```

Has someone an idea, why my .stl models differ from those created with solid works? It would be create, if someone could show me some example code, because i'm not shure if use madcad correctly.

Also, i would like to reference pymadcad in my publications. Is there some prefered citation?

Kind Regards
Stefan

**Comment by jimy-byerley** [comment, marked as answer]
That's an interesting application, I mean machining feature recognition :)
In what precise aspect does the madcad generated meshes differ from the solidworks generated ones ? They look pretty similar on the two pictures you sent

For what regards your use of pymadcad, your code snipet is pretty good, I find nothing wrong or that I would advise to do differently.

I do not have a citation pattern for pymadcad, I don't know the good practices in term of citation so please write what you find appropriate ;)

**Reply by Boehm92** [reply]
Sorry for the late reply. 

After careful evaluation, there were several problems. For starters, there are several mistakes in the benchmark dataset (like random rotations of the cube or wrong depths of the machining features). 
But the main factor is (i assume, cant prove it) the node resolution. For some machining features the node resolution differs from the benchmark dataset. My graph neural network learns the topology of the cad file nodes. So if the resolution differ to much, the graph neural network gives wrong results. With view changes of the network parameters, i got better results. So the problem is not related to madcad and therefore my questions is answered.

I could never found that out without pymadcad. So thousand thanks for the great work and the support  👍


I have on more question:
How can i use the bevel function for only one edge of a cube?
 Maybe i'm confused but i dont get it running. If i only apply it on one edge it throws folloging exception:

```
line 648, in mesh_bevel
r.reverse()
UnboundLocalError: local variable 'r' referenced before assignment
```


My code:
```
   cube = mdc.brick(width=mdc.vec3(2))

    mdc.bevel(
        cube,
        [(1, 2)],  # Edges to smooth
        ("width", 0.3),  # Cutting description, known as 'cutter'
    )
```

**Reply by jimy-byerley** [reply]
I'm glad madcad helped you :)

Regarding your question, your use of bevel is good. But it seems there is a bug with the case of one only edge. I opened #76 for this.
